docs/Threading.md

```diff
@@ -0,0 +1,171 @@
+A thread acquiring a ticket lock will spin in a compare and swap (CAS) loop until it successfully acquires the lock.
+If an lthread yields while holding a ticket lock, there is a chance of deadlock because the next thread to try to acquire the lock will spin in a CAS loop and not yield.
+This may, in turn, prevent the thread holding the lock from being able to run.
+In the simplest case, consider a system running with a single ethread.
+If a thread yields while holding a ticket lock, the next thread that attempts to acquire the same ticket lock will spin and not yield to the scheduler.
+The first thread is never rescheduled and so cannot release the lock.
+Similar situations can occur with multiple ethreads if more than one lthread attempts to acquire ticket locks held by yielded lthreads.
+
+Ticket locks are used to implement some parts of the lthread scheduler.
+This means that there is no way for them to be made to yield in their spin loop.
+Looking at the ticket lock code in isolation may lead you to believe that this limitation could be removed by adding a yield call in the the CAS loop.
+This is not possible because yielding transfers control from the current lthread to the scheduler.
+If the caller is the scheduler then there is no associated lthread that can yield.
+
+#### Lthread futexes
+
+The lthread futex implementation is a simplified version of the Linux `futex` system call.
+A futex is a 32-bit word that can be used to hold state used to implement arbitrary locking primitives.
+On the fast path, all updates to the futex word are done by CPU atomic operations.
+For example, a mutex could be implemented by treating a value of 0 as unlocked and a value of 1 as locked, with a CAS changing the state from 0 to 1 to acquire the lock on the fast path.
+The futex calls handle the slow path, where one or more threads need to wait.
+
+The `enclave_futex_wait` call suspends the calling thread until another thread makes a corresponding `enclave_futex_wake` call for the same futex word address, or until a timeout specified by the caller is reached.
+The futex system is responsible for ensuring that wake-up events are not lost and so acquires a global lock and then checks that the expected value is still present at the memory location.
+This lock ensures that a `enclave_futex_wake` call either happens after the `enclave_futex_wait` has sent the calling thread to sleep or happens before and prevents the thread from sleeping.
+
+The threads suspended by `enclave_futex_wait` are stored in a linked list, threaded through the thread pointers.
+A `enclave_futex_wait` call can return (waking up the sleeping thread) as a result of one of two things.
+Either a `enclave_futex_wake` acquires the lock, finds a waiting thread, and schedules it, or the `futex_tick` call triggers the timeout.
+
+The lthread scheduler calls `futex_tick` in between scheduling threads to wake up any sleeping threads that have timed out.
```

Comment:
How does the `spin` in the scheduler code work?

Reply:
Do you mean `a_spin`?  If so, it's just an x86 `pause` instruction (a hint to superscalar CPUs that they want monitor for cache coherency traffic rather than filling up their load/store pipeline with speculative loads at the same address).

Reply:
No, I meant the `spins` variable. It counts down in the while loop and also gets reset to 0 before the OCALL. The logic beyond the code is not clear to me.